module/interface.py

```diff
@@ -198,7 +202,12 @@ def solve(self):
         Solve QP Problem
         """
         # Solve QP
-        return self._model.solve()
+        results = self._model.solve()
+
+        # TODO(bart): this will be unnecessary when the derivative will be in C
+        self._derivative_cache['results'] = results
```

Comment:
Should we also set this to `None` if `update` is called and then make sure it's populated when the derivative functions are called? This way somebody can't accidentally change the data and then try to differentiate through an old problem.

Also it's ok to not support the setting when the user wants to solve a batch of problems by using `update` on a single OSQP instance and then differentiate through all the solutions, right?

Reply:
You are definitely right. I was thinking that anyway we will not have this issue later when we will port it to C. I have changed to code to set it to `None` when the `update` function gets called.

Reply:
Can we remove the `self._derivative_cache = None` line and add the following code to the bottom of the `update` function? This way the `_derivative_cache` is updated properly and it removes `results` from it, meaning the problem hasn't been solved.

```python
        # update problem data in self._derivative_cache
        if q is not None:
            self._derivative_cache["q"] = q

        if l is not None:
            self._derivative_cache["l"] = l

        if u is not None:
            self._derivative_cache["u"] = u

        if Px is not None:
            if Px_idx.size == 0:
                self._derivative_cache["P"].data = Px
            else:
                self._derivative_cache["P"].data[Px_idx] = Px
        
        if Ax is not None:
            if Ax_idx.size == 0:
                self._derivative_cache["A"].data = Ax
            else:
                self._derivative_cache["A"].data[Ax_idx] = Ax

        # delete results from self._derivative_cache to prohibit
        # taking the derivative of unsolved problems
        if "results" in self._derivative_cache.keys():
            del self._derivative_cache["results"]
```